### PR TITLE
1.0.0 upgrade script upgrade

### PR DESCRIPTION
**Pull request by perk-sumo**
###### Description

Make upgrade script to upgrade only the keys that are used in the client's values.yaml.

###### Testing performed

- [ ] ci/build.sh
- [ ] Redeploy fluentd and fluentd-events pods
- [ ] Confirm events, logs, and metrics are coming in

### CODE REVIEW

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 1]
```diff
@@ -38,17 +40,351 @@ function check_required_command() {
...
+          matchNames:
+            - ${NAMESPACE}
+        selector:
+          matchLabels:
+            app: ${HELM_RELEASE_NAME}-${NAMESPACE}-fluentd-events" | yq m -a -i "new.yaml" -
+
+yq w -i new.yaml -- "prometheus-operator.prometheus.prometheusSpec.containers(name==prometheus-config-reloader).env(name==CHART).valueFrom.configMapKeyRef.key" "fluentdMetrics"
+
+# Check user's image and echo warning if the image has been changed
+readonly USER_VERSION="$(yq r old-values.yaml -- image.tag)"
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:369]: old-values.yaml? 😉 I guess you wanted to have ${OLD_VALUES_YAML}

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 2]
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:2]: add `set -euo pipefail`

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 3]
```diff
@@ -1,5 +1,7 @@
...
 
+#set -x
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:3]: this should go

File: deploy/helm/sumologic/upgrade-1.0.0.sh (hunk 2 above)
Comment (perk-sumo) [reply deploy/helm/sumologic/upgrade-1.0.0.sh:2]: also please `IFS=$'\n\t'`

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 4]
```diff
@@ -38,17 +40,359 @@ function check_required_command() {
...
+sumologic.events.sourceCategory:fluentd.events.sourceCategory
+sumologic.excludeContainerRegex:fluentd.logs.containers.excludeContainerRegex
+sumologic.excludeHostRegex:fluentd.logs.containers.excludeHostRegex
+sumologic.excludeNamespaceRegex:fluentd.logs.containers.excludeNamespaceRegex
+sumologic.excludePodRegex:fluentd.logs.containers.excludePodRegex
+sumologic.fluentd.buffer:fluentd.buffer.type
+sumologic.fluentd.buffer:problems.sumologic.fluentd.buffer
+sumologic.fluentdLogLevel:fluentd.logLevel
+sumologic.flushInterval:fluentd.buffer.flushInterval
+sumologic.k8sMesumologic.k8sMesumologic.k8sMesumologic.k8sMesumologic.k8sMesusMetadataFilter.cacheTtl:fluentd.metadata.cacheTtl
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:77]: please fix that

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 5]
```diff
@@ -38,17 +40,359 @@ function check_required_command() {
...
+sumologic.k8sMetadataFilter.cacheRefresh:fluentd.metadata.cacheRefresh
+sumologic.k8sMetadataFilter.cacheSize:fluentd.metadata.cacheSize
+sumologic.k8sMetadataFilter.cacheTtl:fluentd.metadata.cacheTtl
+sumologic.k8sMetadataFilter.verifySsl:fluentd.logs.containers.k8sMetadataFilter.verifySsl
+sumologic.k8sMetadataFilter.watch:fluentd.logs.containers.k8sMetadataFilter.watch
+sumologic.k8sMetsumologic.k8sMetsumologic.k8sMetsumologic.k8sMetsumoloataFilter.verifySsl
```
Comment (sumo-drosiek) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:83]: review and fix that line

File: deploy/helm/sumologic/upgrade-1.0.0.sh (hunk 4 above)
Comment (sumo-drosiek) [reply deploy/helm/sumologic/upgrade-1.0.0.sh:77]: seems like corrupted line 80, so going to remove that

File: deploy/helm/sumologic/upgrade-1.0.0.sh (hunk 5 above)
Comment (sumo-drosiek) [reply deploy/helm/sumologic/upgrade-1.0.0.sh:83]: Seems like corrupted line 81, so going to remove it

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 6]
```diff
@@ -38,17 +40,359 @@ function check_required_command() {
...
+sumologic.logFormat:fluentd.logs.output.logFormat
+sumologic.numThreads:fluentd.buffer.numThreads
+sumologic.queueChunkLimitSize:fluentd.buffer.queueChunkLimitSize
+sumologic.timestampKey:fluentd.logs.output.timestampKey
+sumologic.totalLimitSize:fluentd.buffer.totalLimitSize
+sumologic.verifySsl:fluentd.verifySsl
+sumologic.watchResourceEventsOverrides:fluentd.events.watchResourceEventsOverrides"
+
+readonly KEY_MAPPINGS_MULTIPLE="
+deployment.affinity:fluentd.logs.statefulset.affinity:fluentd.metrics.statefulset.affinity
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:95]: Please add a warning comment in the ouptut.

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 7]
```diff
@@ -38,17 +40,359 @@ function check_required_command() {
...
+deployment.resources.limits.memory:fluentd.logs.statefulset.resources.limits.memory:fluentd.metrics.statefulset.resources.limits.memory
+deployment.resources.requests.cpu:fluentd.logs.statefulset.resources.requests.cpu:fluentd.metrics.statefulset.resources.requests.cpu
+deployment.resources.requests.memory:fluentd.logs.statefulset.resources.requests.memory:fluentd.metrics.statefulset.resources.requests.memory
+deployment.tolerations:fluentd.logs.statefulset.tolerations:fluentd.metrics.statefulset.tolerations
+sumologic.fluentd.autoscaling.enabled:fluentd.logs.autoscaling.enabled:fluentd.metrics.autoscaling.enabled
+sumologic.fluentd.autoscaling.maxReplicas:fluentd.logs.autoscaling.maxReplicas:fluentd.metrics.autoscaling.maxReplicas
+sumologic.fluentd.autoscaling.minReplicas:fluentd.logs.autoscaling.minReplicas:fluentd.metrics.autoscaling.minReplicas
+sumologic.fluentd.autoscaling.targetCPUUtilizationPercentage:fluentd.logs.autoscaling.targetCPUUtilizationPercentage:fluentd.metrics.autoscaling.targetCPUUtilizationPercentage
+sumologic.sourceCategory:fluentd.logs.containers.sourceCategory:fluentd.logs.kubelet.sourceCategory
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:108]: Those 4 should go into the `fluentd.logs.containers.sourceCategory`

Comment (perk-sumo) [reply deploy/helm/sumologic/upgrade-1.0.0.sh:108]: ... until (and with) this: `sumologic.sourceName`

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 8]
```diff
@@ -38,17 +40,359 @@ function check_required_command() {
...
+  fi
+
+  echo
+done
+
+# Special case for fluentd.events.WatchResourceEventsOverrides
+# as this config is commented out by default but we will write it as empty string
+# which will not work
+if [ "$(yq r $OLD_VALUES_YAML -- sumologic.watchResourceEventsOverrides)" = "" ]; then
+  yq d -i new.yaml -- fluentd.events.watchResourceEventsOverrides
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:179]: This can go.

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 9]
```diff
@@ -38,17 +40,359 @@ function check_required_command() {
...
+elif [ "$(yq r $OLD_VALUES_YAML -- sumologic.addTime)" == "false" ]; then
+  REMOVE="time"
+fi
+
+# Add filter on beginning of current filters
+FILTER="<filter containers.**>
+  @type record_modifier
+  remove_keys $REMOVE
+</filter>
+$(yq r new.yaml -- fluentd.logs.containers.extraFilterPluginConf)"
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:196]: This can go, this key was not use before.

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 10]
```diff
@@ -38,17 +40,359 @@ function check_required_command() {
...
+
+# Apply changes if required
+if [ "$(yq r $OLD_VALUES_YAML -- sumologic.addStream)" == "false" ] || [ "$(yq r $OLD_VALUES_YAML -- sumologic.addTime)" == "false" ]; then
+  yq w -i new.yaml -- fluentd.logs.containers.extraFilterPluginConf "$FILTER"
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:200]: Please add a warning.

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 11]
```diff
@@ -38,17 +40,359 @@ function check_required_command() {
...
+fi
+
+# Keep pre-upgrade hook
+if [[ ! -z "$(yq r new.yaml -- sumologic.setup)" ]]; then
+  yq w -i new.yaml -- 'sumologic.setup.*.annotations[helm.sh/hook]' 'pre-install,pre-upgrade'
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:205]: Please add a warning.

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 12]
```diff
@@ -38,17 +40,359 @@ function check_required_command() {
...
+            else
+                echo "Changes of regex for 'prometheus-operator.prometheus.prometheusSpec.remoteWrite[${i}].writeRelabelConfigs[${j}]' (${metric_name}) detected, please migrate it manually"
+            fi
+        fi
+    fi
+done
+
+# Fix fluent-bit env
+if [[ ! -z "$(yq r new.yaml -- fluent-bit.env)" ]]; then
+  yq w -i new.yaml -- "fluent-bit.env(name==CHART).valueFrom.configMapKeyRef.key" "fluentdLogs"
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:314]: please add a warning

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 13]
```diff
@@ -38,17 +40,359 @@ function check_required_command() {
...
+        namespaceSelector:
+          matchNames:
+            - ${NAMESPACE}
+        selector:
+          matchLabels:
+            app: ${HELM_RELEASE_NAME}-${NAMESPACE}-fluentd-events" | yq m -a -i "new.yaml" -
+fi
+
+if [[ ! -z "$(yq r new.yaml -- prometheus-operator.prometheus.prometheusSpec.containers)" ]]; then
+  yq w -i new.yaml -- "prometheus-operator.prometheus.prometheusSpec.containers(name==prometheus-config-reloader).env(name==CHART).valueFrom.configMapKeyRef.key" "fluentdMetrics"
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:373]: Please add a warning

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 14]
```diff
@@ -38,17 +40,359 @@ function check_required_command() {
...
+fi
+
+# Check user's image and echo warning if the image has been changed
+readonly USER_VERSION="$(yq r ${OLD_VALUES_YAML} -- image.tag)"
+if [[ ! -z "${USER_VERSION}" && "${USER_VERSION}" != "${PREVIOUS_VERSION}" ]]; then
+  echo "You are using incorrect version: ${USER_VERSION}.
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:379]: Please use: You are using unsupported...

**Comment by perk-sumo** [comment]
We need to run it against shellcheck before merge.

**Comment by perk-sumo** [comment]
Also let's ad some proper logging: https://slides.com/perk/how-to-train-your-bash#/41

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 15]
```diff
@@ -27,215 +36,177 @@ For more details, please refer to Migration steps and Changelog here: [link]
...
 check_required_command grep
 check_required_command sed
 check_required_command curl
 
 readonly OLD_VALUES_YAML="$1"
 readonly HELM_RELEASE_NAME="${2:-collection}"
 readonly NAMESPACE="${3:-sumologic}"
+readonly PREVIOUS_VERSION=0.17.2
 
 URL=https://raw.githubusercontent.com/SumoLogic/sumologic-kubernetes-collection/release-v1.0/deploy/helm/sumologic/values.yaml
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:60]: I believe this line can go now. We overwrite it anyway.

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 16]
```diff
@@ -27,215 +36,177 @@ For more details, please refer to Migration steps and Changelog here: [link]
...
+set +e
+IFS=$'\n' read -r -d ' ' -a MAPPINGS <<< "$KEY_MAPPINGS"
+readonly MAPPINGS
+IFS=$'\n' read -r -d ' ' -a MAPPINGS_MULTIPLE <<< "$KEY_MAPPINGS_MULTIPLE"
+readonly MAPPINGS_MULTIPLE
+IFS=$'\n' read -r -d ' ' -a MAPPINGS_EMPTY <<< "$KEY_MAPPINGS_EMPTY"
+readonly MAPPINGS_EMPTY
+set -e
+
+echo > new.yaml
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:138]: Is this needed?

File: deploy/helm/sumologic/upgrade-1.0.0.sh [hunk 17]
```diff
@@ -390,17 +365,25 @@ prometheus-operator:
...
+  warning "Please upgrade to ${PREVIOUS_VERSION} or ensure that new_values.yaml is valid\n"
+fi
 
 # New fluent-bit db path, and account for yq bug that stringifies empty maps
-sed "s/tail-db\/tail-containers-state.db/tail-db\/tail-containers-state-sumo.db/g" new.yaml | \
+info 'Replacing tail-db/tail-containers-state.db to tail-db/tail-containers-state-sumo.db'
+warning 'Please ensure that new fluent-bit configuration is correct\n'
+
+sed 's?tail-db/tail-containers-state.db?tail-db/tail-containers-state-sumo.db?g' new.yaml | \
 sed "s/'{}'/{}/g" > new_values.yaml
```
Comment (perk-sumo) [inline deploy/helm/sumologic/upgrade-1.0.0.sh:387]: I think this is not needed anymore.

Comment (perk-sumo) [reply deploy/helm/sumologic/upgrade-1.0.0.sh]: Nope, it's needed.

**Comment by perk-sumo** [comment]
@sumo-drosiek I've made a couple of refactoring changes: https://github.com/SumoLogic/sumologic-kubernetes-collection/compare/91331e3c26a32118919907faaa2ab1a84fb0cedc...d4f8fd983df9d8ad4fc46428737a49a804fead5b

**Pull request merged**